src/model_theory/semantics.lean

```diff
@@ -273,53 +273,75 @@ lemma realize_cast_le_of_eq {m n : ℕ} (h : m = n) {h' : m ≤ n} {φ : L.bound
+  simp only [cast_le_rfl, cast_refl, order_iso.coe_refl, function.comp.right_id],
+end
+
+lemma realize_map_term_rel_id [L'.Structure M]
+  {ft : ∀ n, L.term (α ⊕ fin n) → L'.term (β ⊕ fin n)}
+  {fr : ∀ n, L.relations n → L'.relations n}
+  {n} {φ : L.bounded_formula α n} {v : α → M} {v' : β → M} {xs : fin n → M}
+  (h1 : ∀ n (t : L.term (α ⊕ fin n)) (xs : fin n → M),
+    (ft n t).realize (sum.elim v' xs) = t.realize (sum.elim v xs))
+  (h2 : ∀ n (R : L.relations n) (x : fin n → M), rel_map (fr n R) x = rel_map R x) :
+  (φ.map_term_rel ft fr (λ _, id)).realize v' xs ↔ φ.realize v xs :=
+begin
+  induction φ with _ _ _ _ _ _ _ _ _ _ _ ih1 ih2 _ _ ih,
+  { refl },
+  { simp [map_term_rel, realize, h1] },
+  { simp [map_term_rel, realize, h1, h2] },
+  { simp [map_term_rel, realize, ih1, ih2], },
+  { simp only [map_term_rel, realize, ih, id.def] },
+end
+
+lemma realize_map_term_rel_add_cast_le [L'.Structure M]
+  {k : ℕ}
+  {ft : ∀ n, L.term (α ⊕ fin n) → L'.term (β ⊕ fin (k + n))}
+  {fr : ∀ n, L.relations n → L'.relations n}
+  {n} {φ : L.bounded_formula α n} {v : α → M} {v' : β → M} {xs : fin (k + n) → M}
+  (h1 : ∀ n (t : L.term (α ⊕ fin n)) (xs' : fin (k + n) → M),
+    (ft n t).realize (sum.elim v' xs') =
+    t.realize (sum.elim v (xs' ∘ fin.nat_add _)))
+  (h2 : ∀ n (R : L.relations n) (x : fin n → M), rel_map (fr n R) x = rel_map R x) :
+  (φ.map_term_rel ft fr (λ n, cast_le (ge_of_eq (add_assoc _ _ _)))).realize v' xs ↔
+    φ.realize v (xs ∘ fin.nat_add _) :=
+begin
+  induction φ with _ _ _ _ _ _ _ _ _ _ _ ih1 ih2 _ _ ih,
+  { refl },
+  { simp [map_term_rel, realize, h1] },
+  { simp [map_term_rel, realize, h1, h2] },
+  { simp [map_term_rel, realize, ih1, ih2], },
+  { simp [map_term_rel, realize, ih] },
 end
 
 lemma realize_relabel {m n : ℕ}
   {φ : L.bounded_formula α n} {g : α → β ⊕ fin m} {v : β → M} {xs : fin (m + n) → M} :
   (φ.relabel g).realize v xs ↔
     φ.realize (sum.elim v (xs ∘ fin.cast_add n) ∘ g) (xs ∘ fin.nat_add m) :=
 begin
-  induction φ with _ _ _ _ _ _ _ _ _ _ _ ih1 ih2 n' _ ih3,
+  rw [relabel],
+  induction φ with _ _ _ _ _ _ _ _ _ _ _ ih1 ih2 _ _ ih,
   { refl },
-  { simp [realize, relabel] },
-  { simp [realize, relabel] },
-  { simp [realize, relabel, ih1, ih2] },
-  { simp only [ih3, realize, relabel],
-    refine forall_congr (λ a, (iff_eq_eq.mpr (congr (congr rfl (congr (congr rfl (congr rfl
-      (funext (λ i, (dif_pos _).trans rfl)))) rfl)) _))),
-    { ext i,
-      by_cases h : i.val < n',
-      { exact (dif_pos (nat.add_lt_add_left h m)).trans (dif_pos h).symm },
-      { exact (dif_neg (λ h', h (nat.lt_of_add_lt_add_left h'))).trans (dif_neg h).symm } } }
+  { simp only [map_term_rel, realize, realize_relabel, sum_elim_comp_relabel_aux], },
+  { simp only [map_term_rel, realize, id.def, realize_relabel, sum_elim_comp_relabel_aux], },
+  { simp [map_term_rel, realize, ih1, ih2] },
+  { simp only [map_term_rel, realize, ih, cast_le_rfl, snoc_comp_nat_add, snoc_comp_cast_add], },
 end
```

Comment:
your auxilliary lemma is not good enough. Use this:
```lean
lemma realize_map_term_rel_add_cast_le [L'.Structure M]
  {k : ℕ}
  {ft : ∀ n, L.term (α ⊕ fin n) → L'.term (β ⊕ fin (k + n))}
  {fr : ∀ n, L.relations n → L'.relations n}
  {n} {φ : L.bounded_formula α n} (v : ∀ {n}, (fin (k + n) → M) → α → M) {v' : β → M} 
  (xs : fin (k + n) → M)
  (h1 : ∀ n (t : L.term (α ⊕ fin n)) (xs' : fin (k + n) → M),
    (ft n t).realize (sum.elim v' xs') =
    t.realize (sum.elim (v xs') (xs' ∘ fin.nat_add _)))
  (h2 : ∀ n (R : L.relations n) (x : fin n → M), rel_map (fr n R) x = rel_map R x) 
  (hv : ∀ n (xs : fin (k + n) → M) (x : M), @v (n+1) (snoc xs x : fin _ → M) = v xs):
  (φ.map_term_rel ft fr (λ n, cast_le (add_assoc _ _ _).symm.le)).realize v' xs ↔
    φ.realize (v xs) (xs ∘ fin.nat_add _) :=
begin
  induction φ with _ _ _ _ _ _ _ _ _ _ _ ih1 ih2 _ _ ih,
  { refl },
  { simp [map_term_rel, realize, h1] },
  { simp [map_term_rel, realize, h1, h2] },
  { simp [map_term_rel, realize, ih1, ih2], },
  { simp [map_term_rel, realize, ih, hv] },
end

lemma realize_relabel {m n : ℕ}
  {φ : L.bounded_formula α n} {g : α → β ⊕ fin m} {v : β → M} {xs : fin (m + n) → M} :
  (φ.relabel g).realize v xs ↔
    φ.realize (sum.elim v (xs ∘ fin.cast_add n) ∘ g) (xs ∘ fin.nat_add m) :=
by rw [relabel, realize_map_term_rel_add_cast_le]; intros; simp
```

Reply:
And you can do even better. The following generalizes my previous lemma, and probably also `realize_lift_at`
```lean
lemma realize_map_term_rel_cast_le [L'.Structure M]
  {g : ℕ → ℕ}
  {ft : ∀ n, L.term (α ⊕ fin n) → L'.term (β ⊕ fin (g n))}
  {fr : ∀ n, L.relations n → L'.relations n}
  {h : ∀ {n}, g (n + 1) ≤ g n + 1}
  {n} {φ : L.bounded_formula α n} (v : ∀ {n}, (fin (g n) → M) → α → M) (s : ∀ {n}, fin n → fin (g n))
  {v' : β → M} {xs : fin (g n) → M} 
  (h1 : ∀ n (t : L.term (α ⊕ fin n)) (xs : fin (g n) → M),
    (ft n t).realize (sum.elim v' xs) = t.realize (sum.elim (v xs) (xs ∘ s)))
  (h2 : ∀ n (R : L.relations n) (x : fin n → M), rel_map (fr n R) x = rel_map R x) 
  (hv : ∀ n (xs : fin (g n) → M) (x : M) {φ : L.bounded_formula α (n+1)}, 
  φ.realize (v (snoc xs x ∘ fin.cast_le h)) ((snoc xs x : fin _ → M) ∘ fin.cast_le h ∘ s) ↔ 
  φ.realize (v xs) (snoc (xs ∘ s) x : fin _ → M)) :
  (φ.map_term_rel ft fr (λ n, cast_le h)).realize v' xs ↔ φ.realize (v xs) (xs ∘ s) :=
begin
  induction φ with _ _ _ _ _ _ _ _ _ _ _ ih1 ih2 _ _ ih,
  { refl },
  { simp [map_term_rel, realize, h1] },
  { simp [map_term_rel, realize, h1, h2] },
  { simp [map_term_rel, realize, ih1, ih2], },
  { simp [map_term_rel, realize, realize_cast_le, ih, ← hv] },
end
```

Reply:
I haven't written a decent `realize_cast_le` yet. Its behavior is unfortunately complicated because of the reverse de Bruijn indexing.

Reply:
Oh, I had one, but I forgot to copy it. I meant the following lemma. Is that supposed to be true, or is this wrong, because of the reversed de Bruijn indexing?

```lean
lemma realize_cast_le {m n : ℕ} (h : m ≤ n) {φ : L.bounded_formula α m}
  {v : α → M} {xs : fin n → M} :
  (φ.cast_le h).realize v xs ↔ φ.realize v (xs ∘ fin.cast_le h)
```

If it's true, I guess the difficulty is getting the induction on `φ` to go through, you'd need to generalize this lemma. You'd need to generalize the right-hand-side to something like `φ.realize v (xs ∘ fin.cast_le h ++ xs')`

Let's do this in a later PR.

bors merge